test_communication/test/test_message_serialization.cpp

```diff
@@ -0,0 +1,223 @@
+  fill_cpp_message(&bounded_array_nested_msg_cpp);
+
+  auto ret = RMW_RET_OK;
+  ret =
+    rmw_serialize(&bounded_array_nested_msg_cpp, message_cpp_typesupport, &raw_message_cpp);
+  EXPECT_EQ(RMW_RET_OK, ret);
+  EXPECT_EQ(76u, raw_message_cpp.buffer_length);
+  print_raw_buffer(raw_message_cpp, "raw message cpp");
+
+  test_msgs::msg::BoundedArrayNested bounded_array_nested_cpp_reverse;
+  ret =
+    rmw_deserialize(&raw_message_cpp, message_cpp_typesupport, &bounded_array_nested_cpp_reverse);
+  EXPECT_EQ(RMW_RET_OK, ret);
+  EXPECT_EQ(true, bounded_array_nested_cpp_reverse.primitive_values[0].bool_value);
+  EXPECT_EQ(255, bounded_array_nested_cpp_reverse.primitive_values[0].byte_value);
+  EXPECT_EQ('k', bounded_array_nested_cpp_reverse.primitive_values[0].char_value);
+  EXPECT_EQ(1, bounded_array_nested_cpp_reverse.primitive_values[0].float32_value);
+  EXPECT_EQ(2, bounded_array_nested_cpp_reverse.primitive_values[0].float64_value);
+  EXPECT_EQ(3, bounded_array_nested_cpp_reverse.primitive_values[0].int8_value);
+  EXPECT_EQ(4u, bounded_array_nested_cpp_reverse.primitive_values[0].uint8_value);
+  EXPECT_EQ(5, bounded_array_nested_cpp_reverse.primitive_values[0].int16_value);
+  EXPECT_EQ(6u, bounded_array_nested_cpp_reverse.primitive_values[0].uint16_value);
+  EXPECT_EQ(7, bounded_array_nested_cpp_reverse.primitive_values[0].int32_value);
+  EXPECT_EQ(8u, bounded_array_nested_cpp_reverse.primitive_values[0].uint32_value);
+  EXPECT_EQ(9, bounded_array_nested_cpp_reverse.primitive_values[0].int64_value);
+  EXPECT_EQ(10u, bounded_array_nested_cpp_reverse.primitive_values[0].uint64_value);
+  EXPECT_STREQ(
+    "hello world", bounded_array_nested_cpp_reverse.primitive_values[0].string_value.c_str());
+
+  rmw_raw_message_fini(&raw_message_cpp);
+}
+
+TEST_F(CLASSNAME(TestMessageSerialization, RMW_IMPLEMENTATION), cdr_integrity) {
+  auto allocator = rcutils_get_default_allocator();
+
+  auto raw_message_c = rmw_get_zero_initialized_raw_message();
+  rmw_initialize_raw_message(&raw_message_c, 0, &allocator);
+
+  auto raw_message_cpp = rmw_get_zero_initialized_raw_message();
+  rmw_initialize_raw_message(&raw_message_cpp, 0, &allocator);
+
+  auto message_c_typesupport = ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, BoundedArrayNested);
+  auto message_cpp_typesupport =
+    rosidl_typesupport_cpp::get_message_type_support_handle<test_msgs::msg::BoundedArrayNested>();
+
+  test_msgs__msg__BoundedArrayNested msg_c;
+  test_msgs__msg__BoundedArrayNested__init(&msg_c);
+  fill_c_message(&msg_c);
+
+  test_msgs::msg::BoundedArrayNested msg_cpp;
+  fill_cpp_message(&msg_cpp);
+
+  auto ret = rmw_serialize(&msg_c, message_c_typesupport, &raw_message_c);
+  ASSERT_EQ(RMW_RET_OK, ret);
+  ret = rmw_serialize(&msg_cpp, message_cpp_typesupport, &raw_message_cpp);
+  EXPECT_EQ(RMW_RET_OK, ret);
+
+  print_raw_buffer(raw_message_c, "raw message c");
+  print_raw_buffer(raw_message_cpp, "raw message cpp");
```

Comment:
Are `raw_message_c.buffer` and `raw_message_cpp.buffer` expected to be identical?